### PR TITLE
feat(block, input-time-picker, notice): adds `open`, `close`, `beforeOpen`, and `beforeClose` events for consistent event handling 

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #5056

## Summary
Implement `openCloseComponent` Interface by using the new `onToggleOpenCloseComponent` pattern. This new pattern is intended to simplify the overall eventing logic.

Components implement the interface to consistently handle emitting `open, beforeOpen, close, beforeClose events`.

- [x] block
- [x] input-time-picker
- [x] notice

### CODE REVIEW

File: packages/calcite-components/src/components/block/block.tsx [hunk 1]
```diff
@@ -104,6 +106,19 @@ export class Block
    */
   @Prop({ reflect: true, mutable: true }) open = false;
 
+  @Watch("open")
+  openHandler(value: boolean): void {
+    if (!this.disabled) {
```
Comment (driskull) [inline packages/calcite-components/src/components/block/block.tsx:111]: nitpick: might be nicer if it was like this...

```
if(this.disabled){
this.open = false;
return;
}
```

File: packages/calcite-components/src/components/block/block.tsx [hunk 2]
```diff
@@ -104,6 +106,19 @@ export class Block
    */
   @Prop({ reflect: true, mutable: true }) open = false;
 
+  @Watch("open")
+  openHandler(value: boolean): void {
```
Comment (driskull) [inline packages/calcite-components/src/components/block/block.tsx:110]: nitpick: should we call `value` => `open`?

File: packages/calcite-components/src/components/block/block.tsx [hunk 3]
```diff
@@ -207,11 +249,17 @@ export class Block
   //
   // --------------------------------------------------------------------------
 
-  /**
-   * Emits when the component's header is clicked.
-   */
-  @Event({ cancelable: false }) calciteBlockToggle: EventEmitter<void>;
```
Comment (driskull) [inline packages/calcite-components/src/components/block/block.tsx:213]: This would be a breaking change right? Do we want to have a breaking change here?

Comment (driskull) [reply packages/calcite-components/src/components/block/block.tsx:213]: Or should we leave it and deprecate it?

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 4]
```diff
@@ -829,7 +868,10 @@ export class InputTimePicker
             scale={this.scale}
             step={this.step}
             // eslint-disable-next-line react/jsx-sort-props
-            ref={this.setCalciteInputEl}
+            ref={(el: HTMLDivElement & HTMLCalciteInputElement) => {
```
Comment (driskull) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:871]: this should just be `HTMLCalciteInputElement` right?

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 5]
```diff
@@ -666,6 +697,10 @@ export class InputTimePicker
     this.popoverEl = el;
   };
 
+  private setTransitionEl = (el: HTMLDivElement): void => {
```
Comment (driskull) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:700]: this should be changed to be `HTMLCalciteInputElement`

File: packages/calcite-components/src/components/notice/notice.tsx [hunk 6]
```diff
@@ -75,6 +73,12 @@ export class Notice
   /** When `true`, the component is visible. */
   @Prop({ reflect: true, mutable: true }) open = false;
 
+  @Watch("open")
+  openHandler(value: boolean): void {
+    onToggleOpenCloseComponent(this);
+    value ? this.open : !this.open;
```
Comment (driskull) [inline packages/calcite-components/src/components/notice/notice.tsx:79]: what is value doing here? Its not updating anywhere.

**Review by driskull — CHANGES_REQUESTED** [review]
Looking good! Just found some things. 🤔

File: packages/calcite-components/src/components/notice/notice.tsx (hunk 6 above)
Comment (Elijbet) [reply packages/calcite-components/src/components/notice/notice.tsx:79]: `openHandler(this.open) is called in the `connectedCallback` for all of these.

Comment (driskull) [reply packages/calcite-components/src/components/notice/notice.tsx:79]: The line of code isn't doing anything

File: packages/calcite-components/src/components/block/block.tsx [hunk 7]
```diff
@@ -149,6 +161,26 @@ export class Block
...
+
+  onBeforeClose(): void {
+    this.calciteBlockBeforeClose.emit();
+  }
+
+  onClose(): void {
+    this.calciteBlockClose.emit();
+  }
+
+  private setTransitionEl = (el: HTMLDivElement): void => {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:180]: This should go under private props.

File: packages/calcite-components/src/components/block/block.tsx [hunk 8]
```diff
@@ -350,6 +402,7 @@ export class Block
             class={CSS.content}
             hidden={!open}
             id={regionId}
+            ref={this.setTransitionEl}
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:405]: Can you add `// eslint-disable-next-line react/jsx-sort-props` before `ref` ([example](https://github.com/Esri/calcite-design-system/blob/f38167bfeaf92557f4f0f9a437b6d4b35b72fc9c/packages/calcite-components/src/components/flow-item/flow-item.tsx#L295))? It's the last prop here, but we need to make sure that it's always set last to ensure props are in sync when `ref`'s callback is invoked.

I'll submit a PR to add an explainer to all of these.

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 9]
```diff
@@ -668,6 +699,7 @@ export class InputTimePicker
 
   private setCalciteInputEl = (el: HTMLCalciteInputElement): void => {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:700]: Function needs to be renamed as it is now setting more than the input element.

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 10]
```diff
@@ -825,11 +861,10 @@ export class InputTimePicker
             onCalciteInputInput={this.calciteInternalInputInputHandler}
             onCalciteInternalInputFocus={this.calciteInternalInputFocusHandler}
             readOnly={readOnly}
+            ref={this.setCalciteInputEl}
             role="combobox"
             scale={this.scale}
             step={this.step}
-            // eslint-disable-next-line react/jsx-sort-props
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:831]: Please restore this (see previous `ref` comment).

File: packages/calcite-components/src/components/notice/notice.tsx [hunk 11]
```diff
@@ -133,6 +136,11 @@ export class Notice
     connectConditionalSlotComponent(this);
     connectLocalized(this);
     connectMessages(this);
+
+    if (this.open) {
+      this.openHandler();
+      onToggleOpenCloseComponent(this);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/notice/notice.tsx:142]: This line isn't needed as `openHandler` already calls this for us.

File: packages/calcite-components/src/components/notice/notice.tsx [hunk 12]
```diff
@@ -168,7 +176,7 @@ export class Notice
     const hasActionEnd = getSlotted(el, SLOTS.actionsEnd);
 
     return (
-      <div class={CSS.container}>
+      <div class={CSS.container} hidden={!open} ref={this.setTransitionEl}>
```
Comment (jcfranco) [inline packages/calcite-components/src/components/notice/notice.tsx:179]: Is the `hidden` change here intentional?

File: packages/calcite-components/src/components/notice/notice.tsx [hunk 13]
```diff
@@ -228,14 +242,34 @@ export class Notice
...
+    this.transitionEl = el;
+  };
+
   //--------------------------------------------------------------------------
   //
   //  Private Methods
   //
   //--------------------------------------------------------------------------
   private close = (): void => {
+    onToggleOpenCloseComponent(this);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/notice/notice.tsx:271]: Should this be called after `open` is modified?

Applies to similar updates.

File: packages/calcite-components/src/utils/openCloseComponent.ts [hunk 14]
```diff
@@ -27,7 +27,7 @@ export interface OpenCloseComponent {
   /**
    * Specifies element that the transition is allowed to emit on.
    */
-  transitionEl: HTMLDivElement;
+  transitionEl: HTMLDivElement | HTMLCalciteInputElement;
```
Comment (jcfranco) [inline packages/calcite-components/src/utils/openCloseComponent.ts:30]: Let's widen the type to `HTMLElement` to allow more elements vs expecting a `div` or `calcite-input`. This should eliminate the need to type cast below.

**Review by jcfranco — CHANGES_REQUESTED** [review]
Looking good, @Elijbet! 🧊😎🧊

**Comment by jcfranco** [comment]
@Elijbet Could you also revisit the PR title to convey what the feature is? Users might not know what "applying the interface" means. On a related note, I think we can drop `openCloseComponent` from the scope since this feature is mostly targeting the other listed components.

File: packages/calcite-components/src/components/block/block.e2e.ts [hunk 15]
```diff
@@ -148,15 +151,15 @@ describe("calcite-block", () => {
 
     await toggle.click();
 
-    expect(toggleSpy).toHaveReceivedEventTimes(1);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.e2e.ts:151]: Let's keep assertions on this event since it's deprecated, not removed.

File: packages/calcite-components/src/components/block/block.tsx [hunk 16]
```diff
@@ -104,6 +106,15 @@ export class Block
    */
   @Prop({ reflect: true, mutable: true }) open = false;
 
+  @Watch("open")
+  openHandler(): void {
+    onToggleOpenCloseComponent(this);
+
+    if (this.disabled) {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:113]: This can be dropped as it's not doing anything.

File: packages/calcite-components/src/components/block/block.tsx [hunk 17]
```diff
@@ -179,6 +210,9 @@ export class Block
     connectInteractive(this);
     connectLocalized(this);
     connectMessages(this);
+    if (this.open) {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:213]: Nitpick: adding a newline here would give it some breathing room between the connect utils and the `open` handling.

File: packages/calcite-components/src/components/block/block.tsx [hunk 18]
```diff
@@ -209,9 +243,22 @@ export class Block
 
   /**
    * Emits when the component's header is clicked.
+   *
+   * @deprecated use `openClose` events instead.
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:247]: +@geospatialem for editorial review.

File: packages/calcite-components/src/components/block/block.tsx [hunk 19]
```diff
@@ -209,9 +243,22 @@ export class Block
...
   @Event({ cancelable: false }) calciteBlockToggle: EventEmitter<void>;
 
+  /** Fires when the component is requested to be closed and before the closing transition begins. */
+  @Event({ cancelable: false }) calciteBlockBeforeClose: EventEmitter<void>;
+
+  /** Fires when the component is closed and animation is complete. */
+  @Event({ cancelable: false }) calciteBlockClose: EventEmitter<void>;
+
+  /** Fires when the component is added to the DOM but not rendered, and before the opening transition begins. */
+  @Event({ cancelable: false }) calciteBlockBeforeOpen: EventEmitter<void>;
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:258]: Nitpick: can the event emitters follow alphabetical order?

Applies to other added emitters.

File: packages/calcite-components/src/components/block/block.tsx [hunk 20]
```diff
@@ -209,9 +243,22 @@ export class Block
...
+
+  /** Fires when the component is open and animation is complete. */
+  @Event({ cancelable: false }) calciteBlockOpen: EventEmitter<void>;
   // --------------------------------------------------------------------------
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:262]: Please add a newline between the emitter and comment divider.

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 21]
```diff
@@ -203,8 +207,8 @@ export class InputTimePicker
 
   @Watch("disabled")
   @Watch("readOnly")
-  handleDisabledAndReadOnlyChange(value: boolean): void {
-    if (!value) {
+  handleDisabledAndReadOnlyChange(open: boolean): void {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:210]: This rename is misleading as the parameter will be whatever `disabled` or `readOnly` was set to (this watch handler is not watching `open`).

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 22]
```diff
@@ -754,6 +792,9 @@ export class InputTimePicker
     connectInteractive(this);
     connectLocalized(this);
 
+    const { open } = this;
+    open && this.openHandler(open);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:796]: Please use an `if` statement for this.

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 23]
```diff
@@ -819,11 +860,10 @@ export class InputTimePicker
             onCalciteInputInput={this.calciteInternalInputInputHandler}
             onCalciteInternalInputFocus={this.calciteInternalInputFocusHandler}
             readOnly={readOnly}
+            ref={this.setInputAndTransitionEl}
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:863]: Can you restore the comment and order of this property? We need it at the end to ensure all props are in sync by the time the ref is called.

File: packages/calcite-components/src/utils/openCloseComponent.ts [hunk 24]
```diff
@@ -166,7 +166,7 @@ export function connectOpenCloseComponent(component: OpenCloseComponent): void {
     const boundOnTransitionEnd: (event: TransitionEvent) => void = transitionEnd.bind(component);
 
     componentToTransitionListeners.set(component, [
-      component.transitionEl,
+      component.transitionEl as HTMLDivElement,
```
Comment (jcfranco) [inline packages/calcite-components/src/utils/openCloseComponent.ts:169]: This is properly typed and `as HTMLDivElement` is no longer needed.

File: packages/calcite-components/src/components/block/block.tsx (hunk 18 above)
Comment (geospatialem) [reply packages/calcite-components/src/components/block/block.tsx:247]: 💥  Looks good!

File: packages/calcite-components/src/components/action-bar/action-bar.tsx [hunk 25]
```diff
@@ -407,8 +407,8 @@ export class ActionBar
     return (
       <calcite-action-group
         class={CSS.actionGroupEnd}
-        label={actionsEndGroupLabel}
         hidden={this.expandDisabled && !(this.hasActionsEnd || this.hasBottomActions)}
+        label={actionsEndGroupLabel}
```
Comment (jcfranco) [inline packages/calcite-components/src/components/action-bar/action-bar.tsx:411]: This seems unrelated. Can we submit this separately?

File: packages/calcite-components/src/components/block/block.e2e.ts [hunk 26]
```diff
@@ -202,15 +208,20 @@ describe("calcite-block", () => {
...
       const block = await page.find("calcite-block");
       const blockToggleSpy = await block.spyOnEvent("calciteBlockToggle");
+      const blockOpenSpy = await block.spyOnEvent("calciteBlockOpen");
+      const blockCloseSpy = await block.spyOnEvent("calciteBlockClose");
 
       await control.press("Space");
       await control.press("Enter");
       await control.click();
+      expect(blockOpenSpy).toHaveReceivedEventTimes(0);
       expect(blockToggleSpy).toHaveReceivedEventTimes(0);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.e2e.ts:218]: For completeness, can we also assert on the close event here?

File: packages/calcite-components/src/components/block/block.tsx [hunk 27]
```diff
@@ -207,8 +238,22 @@ export class Block
...
+  /** Fires when the component is closed and animation is complete. */
+  @Event({ cancelable: false }) calciteBlockClose: EventEmitter<void>;
+
+  /** Fires when the component is open and animation is complete. */
+  @Event({ cancelable: false }) calciteBlockOpen: EventEmitter<void>;
+
   /**
    * Emits when the component's header is clicked.
+   *
+   * @deprecated use `openClose` events instead.
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:256]: Can we explicitly list the events? I'm not sure developers would know what `openClose` events are.

File: packages/calcite-components/src/components/block/block.tsx [hunk 28]
```diff
@@ -350,6 +399,8 @@ export class Block
             class={CSS.content}
             hidden={!open}
             id={regionId}
+            // eslint-disable-next-line react/jsx-sort-props
```
Comment (jcfranco) [inline packages/calcite-components/src/components/block/block.tsx:402]: Can you add the explainer from https://github.com/Esri/calcite-design-system/pull/7584? 

> // eslint-disable-next-line react/jsx-sort-props -- ref should be last so node attrs/props are in sync (see https://github.com/Esri/calcite-design-system/pull/6530)

Applies to other `ref`s.

File: packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx [hunk 29]
```diff
@@ -159,13 +161,15 @@ export class InputTimePicker
   @Prop({ reflect: true, mutable: true }) open = false;
 
   @Watch("open")
-  openHandler(value: boolean): void {
+  openHandler(open: boolean): void {
+    onToggleOpenCloseComponent(this);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/input-time-picker/input-time-picker.tsx:165]: I know we'll no longer be modifying `open` based on other props, but until then, can we move this to the end of the 'closing block' below? Otherwise, events will fire when the component is blocking the component from being opened.

File: packages/calcite-components/src/components/notice/notice.tsx [hunk 30]
```diff
@@ -160,7 +175,7 @@ export class Notice
     const hasActionEnd = getSlotted(el, SLOTS.actionsEnd);
 
     return (
-      <div class={CSS.container}>
+      <div class={CSS.container} ref={this.setTransitionEl}>
```
Comment (jcfranco) [inline packages/calcite-components/src/components/notice/notice.tsx:178]: Can you disable `react/jsx-sort-props` here with the explainer from https://github.com/Esri/calcite-design-system/pull/7584?

**Review by jcfranco — APPROVED** [review]
Awesome stuff, @Elijbet! 🎉

This should be good to merge once comments are addressed.
